Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1541,17 +1541,45 @@ extension BrowserViewController: URLBarDelegate {
     }
 
     func urlBar(urlBar: URLBarView, didSubmitText text: String) {
-        // If we can't make a valid URL, do a search query.
-        // If we still don't have a valid URL, something is broken. Give up.
-        let engine = profile.searchEngines.defaultEngine
-        guard let url = URIFixup.getURL(text) ??
-                        engine.searchURLForQuery(text) else {
-            log.error("Error handling URL entry: \"\(text)\".")
+        let trimmedText = text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())
+        
+        guard let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ") else {
+            submitText(text)
             return
         }
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+        
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) { result in
+                if let urlString = result.successValue {
+                    let engine = self.profile.searchEngines.defaultEngine
```

Comment:
Same here -- arbitrarily using a different engine to morph it with a different template is a red flag that we're doing something wrong, I think. I suspect this would all be simpler if we just inlined the encoding.